huntsman/observatory.py

```diff
@@ -512,3 +378,157 @@ def _create_flat_field_observation(self, alt=None, az=None,
+                                 bias=1000, max_exptime=60., target_adu_percentage=0.5,
+                                 max_num_exposures=10, filter_name=None, min_saturated_exptime=2,
+                                 min_exptime=1, *args, **kwargs):
+        """Take flat fields iteratively by automatically estimating exposure times.
+
+        Args:
+            cameras (dict): Dict of camera name: Camera pairs.
+            filter_names (dict): Dict of filter name for each camera.
+        """
+        target_adu = target_adu_percentage * (min_counts + max_counts)
+        exptimes = {cam_name: [1. * u.second] for cam_name in cameras.keys()}
+
+        # Loop until conditions are met to finish flat-fielding
+        exposure_count = 0
+        while True:
+
+            # Update the exposure count
+            exposure_count += 1
+
+            # Get the FITS headers with a common start time
+            start_time = utils.current_time()
+            fits_headers = self.get_standard_headers(observation=observation)
+            fits_headers['start_time'] = utils.flatten_time(start_time)
+
+            # Take the flat field observations (blocking)
+            next_exptimes = {c: exptimes[c][-1] for c in cameras.keys()}
+            camera_events = self._take_flat_fields(next_exptimes, observation,
+                                                   fits_headers=fits_headers)
+
+            # Check the counts for each image
+            is_saturated = False
+            for cam_name, info in camera_events.items():
+                filename = info['filename']
+                self.logger.debug("Checking counts for {}".format(filename))
+
+                # Load the image we just took
+                try:
+                    data = fits.getdata(filename)
+                except FileNotFoundError:
+                    data = fits.getdata(filename + '.fz')
+
+                # Calculate average counts per pixel
+                mean_counts, _, _ = stats.sigma_clipped_stats(data)
+                mean_counts = mean_counts - bias
+                self.logger.debug("Counts: {}".format(mean_counts))
+                if (mean_counts < min_counts) or (mean_counts > max_counts):
+                    self.logger.debug("Counts outside min/max range, should be discarded.")
+
+                # Check if camera is saturated
+                if mean_counts >= max_counts:
+                    is_saturated = True
+
+                # Get suggested exposure time
+                elapsed_time = (utils.current_time() - start_time).sec
+                exptime = exptimes[cam_name][-1].value * (target_adu / mean_counts)
+                # Account for the sky changing brightness
+                sky_factor = 2.0 ** (elapsed_time / 180.0)
+                if self.past_midnight:
+                    exptime = exptime / sky_factor
+                else:
+                    exptime = exptime * sky_factor
```

Comment:
You may reasonably object on the grounds that it is opaque 'clever code' but you can use `(-1)**self.past_midnight` to have a sign switch depending on whether it is past midnight or not. The boolean value gets coerced into an integer.

```suggestion
                # Account for the sky changing brightness
                exptime *= 2.0**((-1)**self.past_midnight * elapsed_time / 180.0)
```

Reply:
Opaque!(!!!)

Reply:
If you don't like raising `-1` to the power of a boolean property then how about using it as a tuple index? ;-)

```
exptime *= 2**((1, -1)[self.past_midnight] * elapsed_time / 180)
```

BTW, this is Python 3, you don't have to sprinkle `.0` everywhere to avoid accidental integer division any more ;-)